### ISSUE TITLE
Release build fails with SIGSEGV

### ISSUE THREAD

**Issue by svend** [dtolnay/syn#20]
When building with the `cargo build --release`, the build fails with `(signal: 11, SIGSEGV: Invalid memory reference)`. Building with `cargo build` (no `--release` flag) works. Here is the full output, along with the rust version:

```
ubuntu@host:~/syn$ git describe
0.8.0
ubuntu@host:~/syn$ cargo build --release --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling quote v0.2.0
     Running `rustc /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/quote-0.2.0/src/lib.rs --crate-name quote --crate-type lib -C opt-level=3 -C metadata=9442466506b24325 -C extra-filename=-9442466506b24325 --out-dir /home/ubuntu/syn/target/release/deps --emit=dep-info,link -L dependency=/home/ubuntu/syn/target/release/deps --cap-lints allow`
   Compiling syn v0.8.0 (file:///home/ubuntu/syn)
     Running `rustc src/lib.rs --crate-name syn --crate-type lib -C opt-level=3 --cfg feature=\"default\" --cfg feature=\"printing\" --cfg feature=\"parsing\" --cfg feature=\"quote\" -C metadata=977935f812d0e598 --out-dir /home/ubuntu/syn/target/release/deps --emit=dep-info,link -L dependency=/home/ubuntu/syn/target/release/deps --extern quote=/home/ubuntu/syn/target/release/deps/libquote-9442466506b24325.rlib`
error: Could not compile `syn`.

Caused by:
  Process didn't exit successfully: `rustc src/lib.rs --crate-name syn --crate-type lib -C opt-level=3 --cfg feature="default" --cfg feature="printing" --cfg feature="parsing" --cfg feature="quote" -C metadata=977935f812d0e598 --out-dir /home/ubuntu/syn/target/release/deps --emit=dep-info,link -L dependency=/home/ubuntu/syn/target/release/deps --extern quote=/home/ubuntu/syn/target/release/deps/libquote-9442466506b24325.rlib` (signal: 11, SIGSEGV: Invalid memory reference)
ubuntu@host:~/syn$ rustc --version --verbose
rustc 1.12.0 (3191fbae9 2016-09-23)
binary: rustc
commit-hash: 3191fbae9da539442351f883bdabcad0d72efcb6
commit-date: 2016-09-23
host: x86_64-unknown-linux-gnu
release: 1.12.0
```

**Comment by svend** [comment]
Building with rust 1.11.0 works. Let me know if I should file an issue with rust.

```
ubuntu@host:~/syn$ rustc --version --verbose
rustc 1.11.0 (9b21dcd6a 2016-08-15)
binary: rustc
commit-hash: 9b21dcd6a89f38e8ceccb2ede8c9027cb409f6e3
commit-date: 2016-08-15
host: x86_64-unknown-linux-gnu
release: 1.11.0
ubuntu@host:~/syn$ cargo build --release --verbose
    Updating registry `https://github.com/rust-lang/crates.io-index`
   Compiling quote v0.2.0
     Running `rustc /home/ubuntu/.cargo/registry/src/github.com-1ecc6299db9ec823/quote-0.2.0/src/lib.rs --crate-name quote --crate-type lib -C opt-level=3 -C metadata=9442466506b24325 -C extra-filename=-9442466506b24325 --out-dir /home/ubuntu/syn/target/release/deps --emit=dep-info,link -L dependency=/home/ubuntu/syn/target/release/deps -L dependency=/home/ubuntu/syn/target/release/deps --cap-lints allow`
   Compiling syn v0.8.0 (file:///home/ubuntu/syn)
     Running `rustc src/lib.rs --crate-name syn --crate-type lib -C opt-level=3 --cfg feature=\"quote\" --cfg feature=\"default\" --cfg feature=\"printing\" --cfg feature=\"parsing\" --out-dir /home/ubuntu/syn/target/release --emit=dep-info,link -L dependency=/home/ubuntu/syn/target/release -L dependency=/home/ubuntu/syn/target/release/deps --extern quote=/home/ubuntu/syn/target/release/deps/libquote-9442466506b24325.rlib`
```

**Comment by dtolnay** [comment]
Definitely a compiler bug. Looks like it is fixed in beta and nightly so no point filing an issue. We need to work around it here somehow.

It goes away with this commit 3a2ea4bb9149ea65cebb91a3d992fd43c9366315 so I have a lead.

**Comment by svend** [comment]
@dtolnay Thanks. I filed an issue before I read your response. It is here: rust-lang/rust#36875

**Comment by dtolnay** [comment]
Narrowed down further: e8b4e20f2db91710adc175dbf03a79d50f352da9.

**Comment by dtolnay** [comment]
I can't work on this right now but I will do my best to release a workaround later tonight.

**Comment by dtolnay** [comment]
I released a workaround in 0.8.1.

The following code triggered the bug:

```rust
pub fn str_chars(s: &str) -> Vec<char> {
    s.chars().collect()
}
```

The following workaround avoided the bug:

```rust
pub fn str_chars(s: &str) -> Vec<char> {
    let mut result = Vec::new();
    for ch in s.chars() {
        result.push(ch);
    }
    result
}
```
